### PR TITLE
fix [feature] emit an error when trying to call an abstract method #592

### LINKED ISSUE #592: [feature] emit an error when trying to call an abstract method
### Describe the Bug

some edge cases to consider
- abstract static/class methods (I believe this has to be best
- abstract method calls on `super()`

### Sandbox Link

_No response_

### (Only applicable for extension issues) IDE Information

_No response_

### PR DESCRIPTION
**Pull request by asukaminato0721**
fix #592

Added an early diagnostic plus helper methods to emit abstract-method-call whenever an abstract function, staticmethod, or classmethod is invoked directly (or via a class object), preventing previously silent failures while still allowing instance dispatch.

Introduced the new AbstractMethodCall error kind

### CODE REVIEW

File: pyrefly/lib/alt/call.rs [hunk 1]
```diff
@@ -856,6 +868,25 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
...
+                self.is_class_object_type(obj)
+            }
+            _ => false,
+        }
+    }
+
+    fn is_class_object_type(&self, ty: &Type) -> bool {
+        match ty {
+            Type::ClassDef(_) => true,
+            Type::Type(inner) => self.is_class_object_type(inner),
```
Comment (yangdanny97) [inline pyrefly/lib/alt/call.rs]: I don't think we need this case

File: pyrefly/lib/test/abstract_methods.rs [hunk 2]
```diff
@@ -190,6 +190,51 @@ c = Concrete()
 "#,
 );
 
+testcase!(
```
Comment (yangdanny97) [inline pyrefly/lib/test/abstract_methods.rs:193]: I'd like to see a couple of edge cases from the original issue reflected in this test suite:
- If `X.foo()` is illegal, we should also test that calling `foo` on some arbitrary variable of `type[X]` should NOT error unless `X` is final, since it's possible that the variable refers to a subclass of `X` that implements the method.
- We also need a test for calling abstract methods on `super()`

File: pyrefly/lib/alt/call.rs [hunk 3]
```diff
@@ -856,6 +868,25 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
         }
     }
 
+    fn should_error_on_abstract_call(&self, target: &CallTarget) -> bool {
+        match target {
+            CallTarget::Function(_) | CallTarget::FunctionOverload(..) => true,
```
Comment (yangdanny97) [inline pyrefly/lib/alt/call.rs]: Is this code path exercised in a test? Can we have an abstract function/functionOverload be the call target when only methods can be marked abstract?

My hunch is that it's not necessary

File: pyrefly/lib/alt/call.rs [hunk 4]
```diff
@@ -856,6 +868,25 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
...
+            CallTarget::BoundMethod(obj, _) | CallTarget::BoundMethodOverload(obj, ..) => {
+                self.is_class_object_type(obj)
```
Comment (yangdanny97) [inline pyrefly/lib/alt/call.rs]: given there's only one call site let's inline the implementation here

File: pyrefly/lib/alt/call.rs [hunk 5]
```diff
@@ -856,6 +868,25 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
         }
     }
 
+    fn should_error_on_abstract_call(&self, target: &CallTarget) -> bool {
```
Comment (yangdanny97) [inline pyrefly/lib/alt/call.rs]: Can we make a comment here that we're checking against calls to abstract static and class methods from the class, only when the class name is explicitly provided?

File: pyrefly/lib/test/abstract_methods.rs [hunk 6]
```diff
@@ -190,6 +190,84 @@ c = Concrete()
...
+    test_call_abstract_staticmethod_errors,
+    r#"
+from abc import ABC, abstractmethod
+
+class Base(ABC):
+    @staticmethod
+    @abstractmethod
+    def helper() -> None: ...
+
+Base.helper()  # E: Cannot call abstract method `Base.helper`
```
Comment (yangdanny97) [inline pyrefly/lib/test/abstract_methods.rs:217]: We'll still get an error if we have something like

```
x: type[Base] = ...
x.helper()
```

which could potentially lead to false-positives.

This might not be easy to solve, since we don't have access to whether or not the class name is directly referenced when checking static methods (since it's just CallTarget::Function)

The rest of the changes in the PR look good, so I'll merge everything except for the static method checks, and leave that as a TODO for another PR.
